### ISSUE TITLE
[Bug]: JSON request body cannot be accessed if Content-Type header is not application/json

### ISSUE THREAD

**Issue by mrstanb** [nextcloud/server#33973]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

I'm currently trying out a [Nextcloud SCIM application](https://lab.libreho.st/libre.sh/scim/scimserviceprovider) and testing it for basic provisioning via the [SCIM protocol](https://www.simplecloud.info/) together with [Azure AD](https://azure.microsoft.com/en-us/services/active-directory/).

**Sidenote:** I patched the Nextcloud SCIM application, such that it has basic auth disabled and such that it also has a very rudimentary middleware which performs an auth check for a bearer token. This is due to the fact that AzureAD can only supply a bearer token for authentication to a SCIM server, such as the Nextcloud SCIM app.

Basically, AzureAD acts as a client which is making SCIM requests to the Nextcloud app, which itself acts as a server.

As per the SCIM [RFC 7644](https://datatracker.ietf.org/doc/html/rfc7644), clients need to set the `Content-Type` HTTP request header to the value of `application/scim+json`.
As a SCIM client, AzureAD does exactly this, however the Nextcloud SCIM application does not seem to be able to access the HTTP request body (which is formatted as JSON) whenever the `Content-Type` header is set to `application/scim+json` and not to  `application/json`.

### Steps to reproduce

1. Send a POST request with the `Content-Type` header set to `application/scim+json`

```
$ curl -u example:example -H 'Content-Type: application/scim+json' https://<path-to-nextcloud>/index.php/apps/scimserviceprovider/Users -d '{"userName":"someusername"}'
```

2. Result:
* The `userName` POST parameter which is part of the request body is not set to anything in Nextcloud. Calling the method `getParams()` of a class inheriting from `IRequest` does not show the `userName` parameter during debugging either
* If the `Content-Type` header is changed to `application/json`, then the `userName` parameter is set to `someusername` (as expected) and can be also seen when debugging via the `getParams()` method

### Expected behavior

The `userName` POST parameter as part of the POST request body, as well as any other parameters, which are part of the request body, should be accessible and available to use.

### Installation method

Community Web installer on a VPS or web space

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 7.4

### Web server

Other

### Database engine version

MySQL

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost:8888"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "22.2.9.1",
        "overwrite.cli.url": "http:\/\/localhost:8888",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.8.0
  - activity: 2.15.0
  - bruteforcesettings: 2.4.0
  - calendar: 3.3.2
  - circles: 22.3.2
  - cloud_federation_api: 1.5.0
  - comments: 1.12.0
  - contacts: 4.1.1
  - contactsinteraction: 1.3.0
  - dashboard: 7.2.0
  - dav: 1.20.0
  - federatedfilesharing: 1.12.0
  - federation: 1.12.0
  - files: 1.17.0
  - files_pdfviewer: 2.3.1
  - files_rightclick: 1.1.0
  - files_sharing: 1.14.0
  - files_trashbin: 1.12.0
  - files_versions: 1.15.0
  - files_videoplayer: 1.11.0
  - firstrunwizard: 2.11.0
  - logreader: 2.7.0
  - lookup_server_connector: 1.10.0
  - mail: 1.13.4
  - nextcloud_announcements: 1.11.0
  - notifications: 2.10.1
  - oauth2: 1.10.0
  - password_policy: 1.12.0
  - photos: 1.4.0
  - privacy: 1.6.0
  - provisioning_api: 1.12.0
  - recommendations: 1.1.0
  - richdocuments: 4.2.7
  - richdocumentscode: 21.11.402
  - scimserviceprovider: 0.0.1
  - serverinfo: 1.12.0
  - settings: 1.4.0
  - sharebymail: 1.12.0
  - spreed: 12.2.6
  - support: 1.5.0
  - survey_client: 1.10.0
  - systemtags: 1.12.0
  - text: 3.3.1
  - theming: 1.13.0
  - twofactor_backupcodes: 1.11.0
  - updatenotification: 1.12.0
  - user_status: 1.2.0
  - viewer: 1.6.0
  - weather_status: 1.2.0
  - workflowengine: 2.4.0

Disabled:
  - admin_audit
  - encryption
  - files_external
  - user_ldap
```

### Additional info

Web server used is the PHP built-in webserver that can be started from the CLI with `php -S`

**Comment by CarlSchwan** [comment]
I think you could try in lib/private/AppFramework/Http/Request.php to change decodeContent and getContent to do the same thing as for application/json. If it works, please submit a PR :)

**Comment by mrstanb** [comment]
@CarlSchwan Thank you very much for the suggestion!

I have one follow-up question, namely:
Since there are also other possible values for the `Content-Type` header which are different from `application/json`, but still essentially make use of JSON (e.g., [application/jscalendar+json](https://datatracker.ietf.org/doc/html/rfc8984)), would it be a good idea to enable Nextcloud to support all such possible values?

One simple idea that I have, although I'm not sure if it's a viable one, is to just directly check if the request body is a valid JSON via `json_decode`. If yes, just directly parse the JSON and use it.

Moreover, I also read in the HTTP RFC that `Content-Type` as a header SHOULD be included, but it MUST not be necessarily there. This technically would mean that one would not necessarily have to restrict themselves to checking `application/json` as a possible value for a JSON request body. Hence, my idea would be that JSON request bodies could try to be parsed as JSON by Nextcloud without having to restrict oneself to only `application/json` as a possible content type. Would this make sense for Nextcloud?

**Comment by rawtaz** [comment]
> I patched the Nextcloud SCIM application
@mrstanb Are your patches available somewhere? I'm soon about to start poking that SCIM app too, so was just noodling around on the Internet and happened to stumble over this issue :)

**Comment by mrstanb** [comment]
> Are your patches available somewhere? I'm soon about to start poking that SCIM app too, so was just noodling around on the Internet and happened to stumble over this issue :)

@rawtaz Thanks a lot for your interest! We're currently still working on the patches and they'll be available publicly soon. In a way, this publication of the application depends on pull request #34172, since this PR would introduce a change in Nextcloud that we need for the app to essentially work. In the meantime, you can still find the original Nextcloud SCIM app by IndieHosters [here](https://lab.libreho.st/libre.sh/scim/scimserviceprovider)

To give a sense of some of the features that we've added with our patches and extensions to the Nextcloud SCIM app, here's a list with a few examples:
* Support for a subset of all SCIM filtering expressions
* Support for Basic auth, as well as Bearer token auth via JWTs
* More extended support for SCIM users and groups, compared to the original support in the original Nextcloud SCIM app
* Usage of our own SCIM Server PHP library, which you can find [here](https://github.com/audriga/scim-server-php)

**Note:** The SCIM Server PHP Library is also still being developed in parallel with the Nextcloud SCIM app. That's why you can expect it to be updated on GitHub soon as well.

**Comment by rawtaz** [comment]
I really wish the Nextcloud developers could review #34172 and get it going so that your SCIM library is able to actually work as intended. I guess you're completely blocked on #34172?

**Comment by szaimen** [comment]
Hi, please update to 24.0.9 or better 25.0.3 and report back if it fixes the issue. Thank you!

My goal is to add a label like e.g. 25-feedback to this ticket of an up-to-date major Nextcloud version where the bug could be reproduced. However this is not going to work without your help. So thanks for all your effort!

If you don't manage to reproduce the issue in time and the issue gets closed but you can reproduce the issue afterwards, feel free to create a new bug report with up-to-date information by following this link: https://github.com/nextcloud/server/issues/new?assignees=&labels=bug%2C0.+Needs+triage&template=BUG_REPORT.yml&title=%5BBug%5D%3A+

**Comment by mrstanb** [comment]
> Hi, please update to 24.0.9 or better 25.0.3 and report back if it fixes the issue. Thank you!
>
> My goal is to add a label like e.g. 25-feedback to this ticket of an up-to-date major Nextcloud version where the bug could be reproduced. However this is not going to work without your help. So thanks for all your effort!
>
> If you don't manage to reproduce the issue in time and the issue gets closed but you can reproduce the issue afterwards, feel free to create a new bug report with up-to-date information by following this link: https://github.com/nextcloud/server/issues/new?assignees=&labels=bug%2C0.+Needs+triage&template=BUG_REPORT.yml&title=%5BBug%5D%3A+

I tried out Nextcloud v25.0.3 without the changes from #34172 to see how it would behave if I supply anything except `application/json` for when sending a JSON POST or PUT body in an API request against Nextcloud.

I can confirm that without the changes from #34172, the issue persists and only `application/json` is accepted for JSON-formatted request bodies.

**Comment by szaimen** [comment]
So should be fixed with nc26

**Comment by rawtaz** [comment]
A couple of days ago I updated a Nextcloud 24 from 24.0.5 to 24.0.9. I contemplated upgrading to v25 as well, but seeing that there are still apps (not only esoteric third party ones that few people install) that are *still* not marked as compatible with v25, I did not go that route.

It would be good if this change (#34172) could be backported to v25 and v24, because with the current track record of app compatibility it will take a long time before anyone who wants app compatiblity with their Nextcloud version will be able to run v26.

**Comment by joshtrichards** [comment]
#34172 was released with v26.0.0. It wasn't backported to NC25 because no one triggered a backport request. It may have been an oversight or intentional if it was considered too drastic of a change to backport (it doesn't appear that way to me but I didn't look too closely; but I also didn't see anyone bring up the matter on the PR). 

@mrstanb You're welcome to trigger a backport on your own PR using the `/backport stable25` command. It'll generate a PR specific to the latest NC25. It'll still have to get reviewed of course.

NC25 is EOL so soon now at this point...

Anyhow, has anyone confirmed that the core matter discussed in this issue is now solved in >=NC26? If so, let's close this issue out. :)

**Comment by joshtrichards** [comment]
Fixed in #34172

**Comment by mrstanb** [comment]
> @mrstanb You're welcome to trigger a backport on your own PR using the /backport stable25 command. It'll generate a PR specific to the latest NC25. It'll still have to get reviewed of course.

Hi, @joshtrichards, and sorry for the massive delay on my side with that one. I tried to automatically backport #34172 to `stable25` a while ago, but it didn't work out. I now opened a new PR after doing the backport manually (hopefully correctly 😄).

Feel free to check it out: #43538
